ext/standard/random.c

```diff
@@ -0,0 +1,169 @@
+   | license@php.net so we can mail you a copy immediately.               |
+   +----------------------------------------------------------------------+
+   | Authors: Sammy Kaye Powers <me@sammyk.me>                            |
+   +----------------------------------------------------------------------+
+*/
+
+/* $Id$ */
+
+#include <stdlib.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <math.h>
+
+#include "php.h"
+
+#if PHP_WIN32
+# include "win32/winutil.h"
+#endif
+
+static int php_random_bytes(void *bytes, size_t size)
+{
+#if PHP_WIN32
+	/* Defer to CryptGenRandom on Windows */
+	if (php_win32_get_random_bytes(bytes, size) == FAILURE) {
+		php_error_docref(NULL, E_WARNING, "Could not gather sufficient random data");
+		return FAILURE;
+	}
+#else
+#if HAVE_DECL_ARC4RANDOM_BUF
+	arc4random_buf(bytes, size);
+#else
+	int    fd = -1;
+	size_t read_bytes = 0;
+#if HAVE_DEV_ARANDOM
+	fd = open("/dev/arandom", O_RDONLY);
+#else
+#if HAVE_DEV_URANDOM
+	fd = open("/dev/urandom", O_RDONLY);
+#endif // URANDOM
+#endif // ARANDOM
+	if (fd < 0) {
+		php_error_docref(NULL, E_WARNING, "Cannot open source device");
+		return FAILURE;
+	}
+
+	size_t n = 0;
+	while (read_bytes < size) {
+		n = read(fd, bytes + read_bytes, size - read_bytes);
+		if (n < 0) {
+			break;
+		}
+		read_bytes += n;
+	}
+
+	close(fd);
+	if (read_bytes < size) {
+		php_error_docref(NULL, E_WARNING, "Could not gather sufficient random data");
+		return FAILURE;
+	}
+#endif // !ARC4RANDOM_BUF
+#endif // !WIN32
+
+	return SUCCESS;
+}
+
+/* {{{ proto string random_bytes(int length)
+Return an arbitrary length of pseudo-random bytes as binary string */
+PHP_FUNCTION(random_bytes)
+{
+	zend_long size;
+	zend_string *bytes;
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "l", &size) == FAILURE) {
+		return;
+	}
+
+	if (size < 1) {
+		php_error_docref(NULL, E_WARNING, "Length must be greater than 0");
+		RETURN_FALSE;
+	}
+
+	bytes = zend_string_alloc(size, 0);
+
+	if (php_random_bytes(bytes->val, size) == FAILURE) {
+		zend_string_release(bytes);
+		RETURN_FALSE;
+	}
+
+	bytes->val[size] = '\0';
+
+	RETURN_STR(bytes);
+}
+/* }}} */
+
+/* {{{ proto int random_int(int min, int max)
+Return an arbitrary pseudo-random integer */
+PHP_FUNCTION(random_int)
+{
+	zend_long min = ZEND_LONG_MIN;
+	zend_long max = ZEND_LONG_MAX;
+	zend_ulong limit;
+	zend_ulong umax;
+	zend_ulong result;
+
+	if (ZEND_NUM_ARGS() == 1) {
+		php_error_docref(NULL, E_WARNING, "A minimum and maximum value are expected, only minimum given");
+		RETURN_FALSE;
+	}
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "|ll", &min, &max) == FAILURE) {
+		return;
+	}
+
+	if (min >= max) {
+		php_error_docref(NULL, E_WARNING, "Minimum value must be less than the maximum value");
+		RETURN_FALSE;
+	}
+
+	umax = max - min;
+
+	if (php_random_bytes(&result, sizeof(result)) == FAILURE) {
+		RETURN_FALSE;
+	}
+
+	// Special case where no modulus is required
+	if (umax == ZEND_ULONG_MAX) {
```

Comment:
I'm pretty sure this is dead code - an ULONG_MAX is always higher than LONG_MAX, and therefore even if you have `umax = ZEND_LONG_MAX - 0`, it can never be as high ...

Also, for some reason `random_int(0, PHP_INT_MAX)` results in `umax = -1` ...